src/java/org/apache/cassandra/net/Message.java

```diff
@@ -367,8 +368,13 @@ private Header(long id, Verb verb, InetAddressAndPort from, long createdAtNanos,
             this.id = id;
             this.verb = verb;
             this.from = from;
-            this.createdAtNanos = createdAtNanos;
             this.expiresAtNanos = expiresAtNanos;
+            if (verb.isResponse())
+                // Correct createdAtNanos to enforce the contraint, createdAtNanos <= expiresAtNanos
+                this.createdAtNanos = Math.min(createdAtNanos, expiresAtNanos);
```

Comment:
agree. will put more thoughts.

Reply:
Perhaps add a check that expiresAtNanos >= createdAtNanos and either update `responseWith` and `failureResponse` to ensure the time is correct?

Reply:
At the high-level, the coordinator sends request messages to peers and waits for the response messages. The coordinator's waiting time is based on the original request messages' `createdAt` and `expiresAt` (AbstractReadExecutor and AbstractWriteResponseHandler for e.g.). The response message's `createdAt` is not important. 
If looking at the `InboundMessageHandler`, the only field used for expiry check is `header.expiresAtNanos`. The `createdAt` time is not used. 
**Therefore, as long as the `expiresAt` is not changed across nodes, I think it makes no harm.** 

The main benefit of correcting the `createdAt` in response message is that the serialized `VUnsignedInt` field for `createdAt` takes 1 byte instead of 9 bytes in the case of negative values.

We can also revert the constraint if not convinced, since the read/write unsigned vint can apply to negative values in fact with the cost of slightly larger message in some edge cases.

Reply:
I'm unconvinced by this change too - it's accepted and acceptable for rare edge cases to spent 9 bytes encoding a negative value. The cost of those 9 bytes is negligible, and not worth trying to optimise in any way.

Reply:
Reverted the correction to `createdAt`.